packages/shared/src/libs/types/required-permission.ts

```diff
@@ -0,0 +1,8 @@
+import { type PermissionKey as PermissionKeys } from "../enums/permission-key.enum.js";
+import { type ValueOf } from "./value-of.type.js";
+
+type PermissionKey = ValueOf<typeof PermissionKeys>;
+
+type RequiredPermission = PermissionKey | PermissionKey[];
```

Comment:
We don't need this type it's very confusing. We either need to check needed permissions separately or pass a mode as a third argument, which will have values: 'all-of' or 'one-of'
```suggestion
```

Reply:
Agree that this is confising. 

My initial idea was to create two separate functions `checkHasAnyPermissions` and `checkHasAllPermissions`. But the problem with this approach is that it allows only two possible variants: either user should have all of listed permission or at least one of them. But for Project Access Management page the rule is that user should have `Manage User Access` AND eighter `Manage All Projects` OR `View All Projects` permissions, and my initial idea can't handle this.

I'll try to figure out how to solve this.

Reply:
I can't figure out how to deal the case I mentioned without this type. The only idea that comes to my mind is to define this type as followed:
![image](https://github.com/user-attachments/assets/cdc8d1df-9085-4eef-b84d-388451d1218b)

This way it will be used like
![image](https://github.com/user-attachments/assets/8200c5ce-d3cd-415c-83df-208fd33a5501)

and it's much clearer then previous variant, which in this case looks like
![image](https://github.com/user-attachments/assets/c9f6e5ef-af6e-4cbf-aa15-52a0fbec3caa)

![image](https://github.com/user-attachments/assets/87e7a3ab-8dd7-474a-ad08-c361153f7c6d)

cause this (new) way we can clearly see, that user must have all of two permissions: `Manage User Access` and another permission, which could be any of `Manage All Projects`, `View All Projects`

But I'm really not sure if this solution is what you expected to see.

Reply:
How about we rewrite checkHasPermission to something like this? 

```
const checkHasPermission = (
  requiredPermissions: string[],
  permissions: PermissionGetAllItemResponseDto[],
  alternativePermissions?: string[] // Adding an OR condition
): boolean => {
  const hasRequired = requiredPermissions.every((permission) =>
    permissions.some((userPermission) => userPermission.key === permission)
  );

  const hasAlternative = alternativePermissions
    ? alternativePermissions.some((permission) =>
        permissions.some((userPermission) => userPermission.key === permission)
      )
    : true;

  return hasRequired && hasAlternative;
};
```

Reply:
Now it can handle AND + OR combinations

Reply:
Your suggestion less general, but I guess I just overcomplicated the task, and your solution is perfect for now. Thanks a lot! I'll do it like you suggest for now

Reply:
I did it. Hope now it's less confusing

Reply:
Current version is working and is clear, waiting for update to see what you came up with

Reply:
Here is another solution which can fix the problem mentioned in PR title:

![image](https://github.com/user-attachments/assets/62552a35-e0b6-4510-bc34-842cbee56d64)
![image](https://github.com/user-attachments/assets/0d36b280-9ba6-4e3e-a652-d23f92b542a3)

What should I left: current one, or one that on this screenshots?

Reply:
I think this is even simpler and required minimal changes, but what version of CheckHasPermission do you use here?

Reply:
Oh, sorry for confusing. `checkHasPermission` looks like

![image](https://github.com/user-attachments/assets/9da1b2ca-3560-4161-abe2-2f6dc869bf03)